server: create api for table metadata job status

kyle-a-wong opened the pull request:
adds a new api v2 endpoint, GET /api/v2/table_metadata/updatejob.

This new API enables the ability to retrieve the status of the table metadata update job.

Resolves: https://github.com/cockroachdb/cockroach/issues/128896
Epic: [CRDB-37558](https://cockroachlabs.atlassian.net/browse/CRDB-37558)
Release note: None

Note: This PR is stacked on top of #129670, so only the last commit needs review

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/129818)
<!-- Reviewable:end -->

kyle-a-wong committed:
server: create api for table metadata job status

adds a new API v2 endpoint,
`GET /api/v2/table_metadata/updatejob`.

This new API enables the ability to retrieve the status
of the table metadata update job. the job status includes:
 - current_status
 - progress
 - last_start_time
   - the start time of the last job run
 - last_completed_time
   - the time of the last completed job run
 - last_updated_time
   - the time the job progress was last updated

Resolves: #128896
Epic: CRDB-37558
Release note: None

kyle-a-wong reviewed (commented):
Made the changes as we discussed, as well as added some additional information to the job progress detail entity. Let me know what you think of the changes.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129818#-:-O6cBKtw0CM3tag1bsPv:b-v9sn3w)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons, @dhartunian, and @xinhaoz)

___
*[`pkg/server/api_v2_databases_metadata.go` line 728 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/129818#gh-1754889435:-O6cB1gV2DTmVzVr5fuU:b-ivv4hm) ([raw file](https://github.com/cockroachdb/cockroach/blob/961a8b20f033db25524f5eb219f8e1f5c1fc9778/pkg/server/api_v2_databases_metadata.go#L728)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

We won't be able to use these properties unfortunately :(
Can we either get the `running_status` or grab the job progress details direclty somehow? 
https://github.com/cockroachdb/cockroach/pull/129834 is where LastRun was added to the progress details.
</blockquote></details>

I made changes to read this data from the job progress field as we discussed offline.


<!-- Sent from Reviewable.io -->

xinhaoz commented on pkg/server/api_v2_databases_metadata_test.go:
```diff
@@ -495,6 +495,44 @@ func TestGetDBMetadata(t *testing.T) {
+		mdResp := makeApiRequest[tmUpdateJobStatusResponse](t, userClient, ts.AdminURL().WithPath(uri).String())
+		require.Equal(t, "NOT_RUNNING", mdResp.CurrentStatus)
+
+		_, e = conn.Exec(fmt.Sprintf("REVOKE CONNECT ON DATABASE defaultdb FROM %s", sessionUsername.Normalized()))
```
Should this be in an `unauthorized` subtest?

xinhaoz reviewed (approved):
Nice! No blocking comments, feel free to merge after addressing any of the nits.

kyle-a-wong commented on pkg/server/api_v2_databases_metadata.go:
```diff
@@ -657,6 +658,165 @@ func (a *apiV2Server) getDBMetadata(
+	)
+`, sqlUserStr, sqlUserStr)
+
+	it, err := a.sqlServer.internalExecutor.QueryIteratorEx(
```
Done

kyle-a-wong commented on pkg/server/api_v2_databases_metadata.go:
```diff
@@ -657,6 +658,165 @@ func (a *apiV2Server) getDBMetadata(
+	query := safesql.NewQuery()
+	sqlUserStr := sqlUser.Normalized()
+	query.Append(`
+	SELECT count(*)
+	FROM (
+	  SELECT 1 FROM system.role_members WHERE member = $ AND role = 'admin'
+		UNION
+		SELECT 1 
+		FROM "".crdb_internal.cluster_database_privileges cdp
+	 	WHERE cdp.grantee = $ 
+	 	AND cdp.privilege_type = 'CONNECT' 
```
oh hm interesting. I can't remember if i looked into that or not / why i decided to use cluster database privileges. I'll look into it again, but probably wont block this PR on it

kyle-a-wong commented on pkg/server/api_v2_databases_metadata.go:
```diff
@@ -657,6 +658,165 @@ func (a *apiV2Server) getDBMetadata(
+		return jobStatus, err
+	}
+
+	if ok {
```
Changed to QueryRow so this is no longer needed.

kyle-a-wong reviewed (commented):
Addressed most of the comments. Wasn't sure what you meant by:
`I don't think there would be but is there any consequence to setting this while executing the txn?` So I didn't make any changes around it.
